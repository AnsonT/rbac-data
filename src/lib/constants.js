exports.ROOT_TENANT = '00000000-0000-0000-0000-000000000000'
exports.GLOBAL_TENANT = 'ffffffff-ffff-ffff-ffff-ffffffffffff'
exports.SUPERUSER_ROLE = '00000000-0000-0000-0000-000000000000'
exports.ADMIN_ROLE = '00000000-0000-0000-0000-000000000001'
exports.USER_ROLE = '00000000-0000-0000-0000-000000000002'
