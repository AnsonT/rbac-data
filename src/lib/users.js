import uuid from 'uuid/v4'
import _ from 'lodash-uuid'
import emailValidator from 'email-validator'
import { InvalidParameterError } from './errors'
import { ROOT_TENANT } from './constants'

export async function createUser (tx, { tenantId, userName, email, emailVerifiedAt, needNewPassword } = {}) {
  if (_.isUuid(userName)) {
    throw new InvalidParameterError({ message: 'Invalid parameter: userName' })
  }
  if (!emailValidator.validate(email)) {
    throw new InvalidParameterError({ message: 'Invalid parameter: email' })
  }
  userName = userName.toLowerCase()
  email = email.toLowerCase()
  const userId = uuid()
  const resp = await tx
    .insert({
      userId,
      userName,
      email,
      emailVerifiedAt,
      needNewPassword
    })
    .into('users')
  if (resp.rowCount === 1) {
    return { userId, userName }
  }
}

export async function listUsers (tx, { tenantId = ROOT_TENANT, offset = 0, limit = 20 }) {}

/**
 * Return a user by userId
 * @param {object} tx - Knex transaction object
 * @param {string} userId - the userId of the user
 */
export async function getUserById (tx, userId) {
  return tx
    .select()
    .from('users')
    .where({ userId })
    .first()
}

/**
 * Return a user based on userName or email (userName or email must be specified)
 * @param {object} tx - Knex transaction object
 * @param {object} where - the query parameters for the user
 * @param {string} where.tenantId - tenant realm of user, defaults to ROOT_TENANT
 * @param {string} where.userName - user name
 * @param {string} where.email - email
 */
export async function getUserByNameOrEmail (tx, { tenantId = ROOT_TENANT, userName, email }) {
  if (!userName && !email) {
    throw new InvalidParameterError({ message: 'userName or email must be specified' })
  }
  userName = userName && userName.toLowerCase()
  email = email && email.toLowerCase()
  tx = tx
    .select()
    .from('users')
    .where({ tenantId })
  if (userName) { tx = tx.where({ userName }) }
  if (email) { tx = tx.where({ email }) }
  return tx.first()
}

export async function removeUser (tx, userId) {}

export async function updateUser (tx, userId, { userName, email, emailVerifiedAt, needNewPassword }) {}
