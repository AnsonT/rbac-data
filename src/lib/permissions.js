import _ from 'lodash'
import uuid from 'uuid/v4'
import { InvalidParameterError } from './errors'
import { GLOBAL_TENANT, ROOT_TENANT } from './constants'
import { getUserByName } from './users'

function validatePermission (permission, global) {
  if (!permission) {
    return undefined
  }
  if (global && permission[0] !== '_') {
    throw new InvalidParameterError({ message: 'Global permissions must start with _' })
  }
  if (!global && permission[0] === '_') {
    throw new InvalidParameterError({ message: 'Local permissions cannot start with _' })
  }
  return permission
}

export async function createPermission (tx, { tenantId, permission, description, global, createdBy }) {
  const permissionId = uuid()
  if (permission.startsWith('_')) {
    global = true
  }
  if (tenantId && global) {
    throw new InvalidParameterError({ message: 'Cannot specify both tenantId and global: true' })
  }
  if (!global && !tenantId) {
    throw new InvalidParameterError({ message: 'Must specify tenantId, or set global: true' })
  }
  global = !tenantId
  tenantId = global ? GLOBAL_TENANT : tenantId
  permission = validatePermission(permission, global)
  const values = {
    permissionId,
    tenantId,
    permission,
    description,
    global,
    createdBy
  }
  await tx
    .insert(values)
    .into('permissions')
  return { permissionId, tenantId, permission, global }
}

export async function getPermissionById (tx, permissionId) {
  return tx
    .select()
    .where({ permissionId })
    .from('permissions')
    .first()
}

export async function getPermissionByName (tx, { tenantId, permission }) {
  if (tenantId) {
    return tx
      .select()
      .where({ permission })
      .where((builder) => builder.where({ tenantId }).orWhere({ tenantId: GLOBAL_TENANT }))
      .orderBy('global', 'asc')
      .from('permissions')
      .first()
  } else {
    return tx
      .select()
      .where({ permission })
      .where({ global: true })
      .from('permissions')
      .first()
  }
}

export async function listPermissions (tx, { tenantId = ROOT_TENANT } = {}) {
  return tx
    .select()
    .where({ tenantId })
    .orWhere({ tenantId: GLOBAL_TENANT })
    .orderBy('global', 'asc')
    .from('permissions')
}

export async function removePermission (tx, permissionId) {
  const count = await tx
    .from('permissions')
    .where({ permissionId })
    .whereNot({ canBeDeleted: false })
    .del()
  return { permissionId, count }
}

export async function updatePermission (tx, permissionId, { permission, description, modifiedBy }) {
  const { global } = await getPermissionById(tx, permissionId)
  permission = validatePermission(permission, global)
  const modifiedAt = tx.fn.now()
  const values = {
    permission,
    description,
    modifiedAt,
    modifiedBy
  }
  const count = await tx
    .where({ permissionId })
    .update(values)
    .into('permissions')
  return { count }
}

export async function listUserPermissions (tx, userId) {
  const permissions = await tx
    .select(['p.permissionId', 'p.permission', 'p.description', 'p.global', 'rp.denied'])
    .from('permissions as p')
    .join('rolesPermissions as rp', 'rp.permissionId', 'p.permissionId')
    .join('roles as r', 'r.roleId', 'rp.roleId')
    .join('usersRoles as ur', 'ur.roleId', 'r.roleId')
    .where('ur.userId', '=', userId)

  const allow = _.map(_.filter(permissions, { denied: false }), p => p.permission)
  const deny = _.map(_.filter(permissions, { denied: true }), p => p.permission)
  return {
    allow,
    deny,
    grants: _.filter(allow, a => !_.includes(deny, a)),
    permissions
  }
}

export function checkPermissions (requiredPermissions, grants, { any = false } = {}) {
  if (!grants || !_.isArray(grants)) {
    throw new InvalidParameterError({ message: 'grants must be an array' })
  }
  if (_.isString(requiredPermissions)) {
    return _.includes(grants, requiredPermissions)
  }
  if (_.isArray(requiredPermissions)) {
    const check = _.curry(checkPermissions)(_, grants)
    return any
      ? _.some(requiredPermissions, check)
      : _.every(requiredPermissions, check)
  }
  const allRequired = _.get(requiredPermissions, 'all')
  const anyRequired = _.get(requiredPermissions, 'any')
  if (allRequired && anyRequired) {
    throw new InvalidParameterError({ message: 'requirePermissions cannot have all and any in the same condition' })
  }
  if (allRequired) {
    return checkPermissions(allRequired, grants)
  }
  if (anyRequired) {
    return checkPermissions(anyRequired, grants, { any: true })
  }
  return false
}

export async function checkUserPermissionsByName (tx, requiredPermissions, { tenantId = ROOT_TENANT, userName }) {
  const { userId } = await getUserByName(tx, { tenantId, userName })
  const { grants } = await listUserPermissions(tx, userId)
  return checkPermissions(requiredPermissions, grants)
}

export async function listRolePermissionsByName (tx, { tenantId = ROOT_TENANT, roleName }) {
  return tx
    .select()
    .from('permissions as p')
    .join('rolesPermissions as rp', 'rp.permissionId', 'p.permissionId')
    .join('roles as r', 'r.roleId', 'rp.roleId')
    .where('r.roleName', '=', roleName)
    .where('r.tenantId', '=', tenantId)
}

export async function listRolePermissionsById (tx, roleId) {
  return tx
    .select()
    .from('permissions as p')
    .join('rolesPermissions as rp', 'rp.permissionId', 'p.permissionId')
    .join('roles as r', 'r.roleId', 'rp.roleId')
    .where('r.roleId', '=', roleId)
}
