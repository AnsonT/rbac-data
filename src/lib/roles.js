import { ROOT_TENANT, GLOBAL_TENANT } from './constants'
import uuid from 'uuid/v4'
import _ from 'lodash-uuid'
import { InvalidParameterError } from './errors'
import { getUserById, getUserByName } from './users'

function validateRoleName (roleName) {
  if (!roleName) { return undefined }
  if (_.isUuid(roleName)) {
    throw new InvalidParameterError({ message: 'Invalid parameter: roleName' })
  }
  return roleName
}

export async function createRole (tx, { tenantId = ROOT_TENANT, roleName, description } = {}) {
  const roleId = uuid()
  if (!roleName) {
    throw new InvalidParameterError({ message: 'roleName is required' })
  }
  if (tenantId === GLOBAL_TENANT) {
    throw new InvalidParameterError({ message: 'global roles are supported' })
  }
  roleName = validateRoleName(roleName)
  const values = {
    roleId,
    tenantId,
    roleName,
    description
  }
  await tx
    .insert(values)
    .into('roles')
  return { roleId, tenantId, roleName }
}

export async function getRoleById (tx, roleId) {
  return tx
    .select()
    .where({ roleId })
    .from('roles')
    .first()
}

export async function getRoleByName (tx, { tenantId = ROOT_TENANT, roleName } = {}) {
  if (!roleName) {
    throw new InvalidParameterError({ message: 'roleName is required' })
  }
  return tx
    .select()
    .where({ tenantId, roleName })
    .from('roles')
    .first()
}

export async function listRoles (tx, { tenantId = ROOT_TENANT, offset = 0, limit = 20 } = {}) {
  return tx
    .select()
    .where({ tenantId })
    .offset(offset)
    .limit(limit)
    .from('roles')
}

export async function removeRole (tx, roleId) {
  const count = await tx
    .from('roles')
    .where({ roleId })
    .whereNot({ canBeDeleted: false })
    .del()
  return { roleId, count }
}

export async function updateRole (tx, roleId, { roleName, description }) {
  const modifiedAt = tx.fn.now()
  roleName = validateRoleName(roleName)
  const values = {
    roleName,
    description,
    modifiedAt
  }
  const count = await tx
    .where({ roleId })
    .update(values)
    .into('roles')
  return { count }
}

export async function assignUserRoleById (tx, { userId, roleId }) {
  const { tenantId: userTenantId } = await getUserById(tx, userId)
  const { tenantId: roleTenantId } = await getRoleById(tx, roleId)
  if (userTenantId !== roleTenantId) {
    throw new InvalidParameterError({ message: 'Role and User from different tenants' })
  }
  await tx
    .insert({ userId, roleId })
    .into('usersRoles')
  return { userId, roleId }
}

export async function assignUserRoleByName (tx, { tenantId = ROOT_TENANT, userName, roleName } = {}) {
  const { userId } = getUserByName(tx, { tenantId, userName })
  const { roleId } = getRoleByName(tx, { tenantId, roleName })
  await tx
    .insert({ userId, roleId })
    .into('usersRoles')
  return { userId, roleId }
}

export async function removeUserRole (tx, userId, roleId) {
  const count = await tx
    .from('usersRoles')
    .del()
    .where({ userId, roleId })
  return { userId, roleId, count }
}

export async function grantRolePermissionById (tx, roleId, permissionId) {
  await tx
    .insert({ roleId, permissionId, denied: false })
    .into('rolesPermissions')
  return { roleId, permissionId }
}

export async function denyRolePermissionById (tx, roleId, permissionId) {
  await tx
    .insert({ roleId, permissionId, denied: true })
    .into('rolesPermissions')
  return { roleId, permissionId }
}

export async function listUserRoles (tx, userId) {
  const roles = await tx
    .select()
    .from('usersRoles')
    .join('roles', 'usersRoles.roleId', 'roles.roleId')
    .where({ userId })
  return roles
}

export async function setUserRolesByName (tx, { tenantId = ROOT_TENANT, userName, roleNames }) {
  const { userId } = await getUserByName(tx, { tenantId, userName }) || {}
  if (!userId) {
    throw new InvalidParameterError({ message: `Invalid userName: "${userName}"` })
  }
  await tx
    .from('usersRoles')
    .del()
    .where({ userId })
  for (const roleName of roleNames) {
    const { roleId } = await getRoleByName(tx, { tenantId, roleName }) || {}
    if (!roleId) {
      throw new InvalidParameterError({ message: `Invalid roleName: "${roleName}"` })
    }
    await assignUserRoleById(tx, { userId, roleId })
  }
  return true
}

export async function setUserRolesById (tx, userId, roleIds) {
  await tx
    .from('usersRoles')
    .del()
    .where({ userId })
  for (const roleId of roleIds) {
    await assignUserRoleById(tx, { userId, roleId })
  }
  return true
}
