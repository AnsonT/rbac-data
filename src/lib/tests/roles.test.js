import _ from 'lodash'
import { UnitTestDb } from './db'
import { createRole, getRoleById, getRoleByName, listRoles, removeRole, updateRole, assignUserRoleById, listUserRoles, listUserPermissions, allowPermissionInRoleById, denyPermissionInRoleById } from '../roles'
import { SUPERUSER_ROLE, ROOT_TENANT } from '../constants'
import uuid from 'uuid/v4'
import { createUser, getUserByName } from '../users'
import { createPermission } from '../permissions'

const user1 = {
  userName: 'User1',
  email: 'user1@example.localhost',
  password: 'password'
}
const role1 = {
  roleName: 'role1',
  description: 'role 1'
}

const role2 = {
  roleName: 'role2',
  description: 'role 2'
}

const role3 = {
  roleName: 'role3',
  description: 'role 3'
}

const permission1 = {
  permission: 'perm1',
  description: 'root permission 1',
  tenantId: ROOT_TENANT
}

const permission2 = {
  permission: 'perm2',
  description: 'root permission 2',
  tenantId: ROOT_TENANT
}

const permission3 = {
  permission: '_perm3',
  description: 'root permission 3',
  global: true
}

describe('Roles Tests', () => {
  const db = new UnitTestDb()
  let knex = null

  beforeAll(async (done) => {
    await db.initialize('permissions')
    knex = db.knex
    done()
  })
  afterAll(async (done) => {
    await db.cleanup()
    done()
  })
  it('createRole succeeds', async (done) => {
    const { roleId } = await createRole(knex, role1)
    expect(roleId).not.toBeNull()
    const { roleName } = await getRoleById(knex, roleId)
    expect(roleName).toBe(role1.roleName)
    await createRole(knex, role2)
    await createRole(knex, role3)
    done()
  })
  it('getRoleByName succeeds', async (done) => {
    const { roleName } = await getRoleByName(knex, { roleName: role1.roleName })
    expect(roleName).toBe(role1.roleName)
    done()
  })
  it('listRoles succeeds', async (done) => {
    const omitMutableFields = (users) => _.map(users, user => _.omit(user, ['roleId', 'createdAt', 'modifiedAt']))
    const roles = await listRoles(knex)
    expect(omitMutableFields(roles)).toMatchSnapshot()
    done()
  })
  it('removeRole succeeds', async (done) => {
    const { roleId } = await getRoleByName(knex, { roleName: role2.roleName })
    const { count } = await removeRole(knex, roleId)
    expect(count).toBe(1)
    const roles = await listRoles(knex)
    expect(roles.length).toMatchSnapshot()
    done()
  })
  it('removeRole superuser fails', async (done) => {
    const { count } = await removeRole(knex, SUPERUSER_ROLE)
    expect(count).toBe(0)
    done()
  })
  it('updateRole succeeds', async (done) => {
    const { roleId } = await getRoleByName(knex, { roleName: role1.roleName })
    const { count } = await updateRole(knex, roleId, { description: 'changed description' })
    expect(count).toBe(1)
    const { description } = await getRoleById(knex, roleId)
    expect(description).toBe('changed description')
    done()
  })
  it('updateRole invalid roleName fails', async (done) => {
    const { roleId } = await getRoleByName(knex, { roleName: role1.roleName })
    await expect(updateRole(knex, roleId, { roleName: uuid() })).rejects.toThrow()
    done()
  })
  it('addUserRole by Id succeeds', async (done) => {
    const { userId } = await createUser(knex, user1)
    const { roleId: role1Id } = await getRoleByName(knex, { roleName: role1.roleName })
    const { roleId: role3Id } = await getRoleByName(knex, { roleName: role3.roleName })
    await assignUserRoleById(knex, {
      userId,
      roleId: role1Id
    })
    await assignUserRoleById(knex, {
      userId,
      roleId: role3Id
    })
    const omitMutableFields = (users) =>
      _.map(users, user => _.omit(user,
        ['roleId', 'userId', 'createdAt', 'modifiedAt']))
    const roles = await listUserRoles(knex, userId)
    expect(omitMutableFields(roles)).toMatchSnapshot()
    done()
  })
  it('addRolePermissions succeeds', async (done) => {
    const { roleId } = await getRoleByName(knex, { roleName: role1.roleName })
    const { permissionId: permission1Id } = await createPermission(knex, permission1)
    const { permissionId: permission2Id } = await createPermission(knex, permission2)
    const { permissionId: permission3Id } = await createPermission(knex, permission3)

    await allowPermissionInRoleById(knex, roleId, permission1Id)
    await allowPermissionInRoleById(knex, roleId, permission2Id)
    await denyPermissionInRoleById(knex, roleId, permission3Id)
    done()
  })
  it('listUserPermissions succeeds', async (done) => {
    const omitMutableFields = (users) =>
      _.map(users, user => _.omit(user,
        ['permissionId']))

    const { userId } = await getUserByName(knex, user1)
    const { allowed, denied } = await listUserPermissions(knex, userId)
    expect(omitMutableFields(allowed)).toMatchSnapshot()
    expect(omitMutableFields(denied)).toMatchSnapshot()
    done()
  })
})
