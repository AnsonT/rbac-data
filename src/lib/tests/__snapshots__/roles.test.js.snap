// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Roles Tests addUserRole by Id succeeds 1`] = `
Array [
  Object {
    "canBeDeleted": true,
    "description": "role 3",
    "roleName": "role3",
    "tenantId": "00000000-0000-0000-0000-000000000000",
  },
  Object {
    "canBeDeleted": true,
    "description": "changed description",
    "roleName": "role1",
    "tenantId": "00000000-0000-0000-0000-000000000000",
  },
]
`;

exports[`Roles Tests listRoles succeeds 1`] = `
Array [
  Object {
    "canBeDeleted": false,
    "description": "root superuser",
    "roleName": "superuser",
    "tenantId": "00000000-0000-0000-0000-000000000000",
  },
  Object {
    "canBeDeleted": true,
    "description": "role 1",
    "roleName": "role1",
    "tenantId": "00000000-0000-0000-0000-000000000000",
  },
  Object {
    "canBeDeleted": true,
    "description": "role 2",
    "roleName": "role2",
    "tenantId": "00000000-0000-0000-0000-000000000000",
  },
  Object {
    "canBeDeleted": true,
    "description": "role 3",
    "roleName": "role3",
    "tenantId": "00000000-0000-0000-0000-000000000000",
  },
]
`;

exports[`Roles Tests listUserPermissions succeeds 1`] = `
Array [
  Object {
    "denied": false,
    "description": "root permission 1",
    "global": false,
    "permission": "perm1",
  },
  Object {
    "denied": false,
    "description": "root permission 2",
    "global": false,
    "permission": "perm2",
  },
]
`;

exports[`Roles Tests listUserPermissions succeeds 2`] = `
Array [
  Object {
    "denied": true,
    "description": "root permission 3",
    "global": true,
    "permission": "_perm3",
  },
]
`;

exports[`Roles Tests removeRole succeeds 1`] = `3`;
